teams/vsa/engineering/forms-system/custom_component_multiple.md

```diff
@@ -0,0 +1,138 @@
+
+## The Problem:
+Using a custom component requires that you pass in a schema for that component.  It essentially maps a small slice of form state to the component.  Because the same component is used twice but we are passing in a different schema, each of those components, each of those only has access to its own individual schema/form state. (Read [Background: The Original Problem](#background-the-original-problem) for further details.)
+
+## The Solution/Hack:
+Take the custom React widget and connect it to Redux.  After its connected to Redux, map state to the entire form state so it can see the form state of other components.  From there, we need to be able to update form state for events we've added to the custom widget.  I am leveraging *setData* located in [platform/forms-system/src/js/actions.js](https://github.com/department-of-veterans-affairs/vets-website/blob/master/src/platform/forms-system/src/js/actions.js) - I basically looked at what data the form system would send it naturally and mimicked it in my own *onChange* event here: https://github.com/department-of-veterans-affairs/vets-website/pull/12295/files#diff-e840e38183aa55fb387f1af118cd752bR92
+
+*[src/applications/disability-benefits/2346/components/ReviewCardField.jsx](https://github.com/department-of-veterans-affairs/vets-website/blob/master/src/applications/disability-benefits/2346/components/ReviewCardField.jsx)*
+
+```
+import { connect } from 'react-redux';
+import { setData } from 'platform/forms-system/src/js/actions';
+
+. . .
+
+  onChange = (field, data) => {
+    const newData = set(field, data, this.props.data);
+    return this.props.setData(newData);
+  };
+
+. . .
+
+  <input
+    id={this.props.name}
+    type="radio"
+    checked={this.props.currentAddress === this.props.name}
+    onChange={() =>
+    this.onChange('currentAddress', this.props.name)
+    }
+  />
+
+. . .
+
+const mapStateToProps = state => ({
+  data: state.form?.data,
+  currentAddress: state.form?.data?.currentAddress,
+});
+const mapDispatchToProps = {
+  setData,
+};
+export default connect(
+  mapStateToProps,
+  mapDispatchToProps,
+)(ReviewCardField);
+
+  ```
+
+## Side-Effect: Extra Component
+You still need to pass in a schema that accounts for the data you want to modify with the form system.  The form system will then add its own component for that data.  So you end up with an extra component.  We need to hide it!  The original way to do this was to leverage a hideIf param that exits in the uiSchema.  There is, however, a gotcha with this.  When you hide a component on the form, it removes its schema thus wiping out its form state.
+
+*[src/platform/forms-system/src/js/state/helpers.js](https://github.com/department-of-veterans-affairs/vets-website/blob/master/src/platform/forms-system/src/js/state/helpers.js)*
+```
+export function expandArrayPages(pageList, data) {
+. . .
+     } else if (nextPage.arrayPath !== lastArrayPath && !!arrayPages.length) {
```

Comment:
@bradl3yC is this really the right line of code?

Reply:
@short000 this is the block that removes it:

```export function removeHiddenData(schema, data) {
  // null is necessary here because Rails 4 will convert empty arrays to null
  // In the forms, there's no difference between an empty array and null or undefined
  if (isHiddenField(schema) || typeof data === 'undefined' || data === null) {
    return undefined;
  }

  if (schema.type === 'object') {
    return Object.keys(data).reduce((current, next) => {
      if (typeof data[next] !== 'undefined' && schema.properties[next]) {
        const nextData = removeHiddenData(schema.properties[next], data[next]);

        // if the data was removed, then just unset it
        if (typeof nextData === 'undefined') {
          return _.unset(next, current);
        }

        // if data was updated (like a nested prop was removed), update it
        if (nextData !== data[next]) {
          return _.set(next, nextData, current);
        }
      }

      return current;
    }, data);
  }

  if (schema.type === 'array') {
    const newItems = data.map((item, index) =>
      removeHiddenData(schema.items[index], item),
    );

    if (newItems.some((newItem, idx) => newItem !== data[idx])) {
      return newItems;
    }

    return data;
  }

  return data;
}```

Its in src/platform/forms-system/src/js/state/helpers.js line 566